src/unix/mmap.c

```diff
@@ -192,15 +192,59 @@ sysreturn mremap(void *old_address, u64 old_size, u64 new_size, int flags, void
     return sysreturn_from_pointer(vnew);
 }
 
-static sysreturn mincore(void *addr, u64 length, u8 *vec)
+static CLOSURE_3_3(mincore_fill_vec, boolean, u64, u64, u8 *, int, u64, u64 *);
+boolean mincore_fill_vec(u64 base, u64 nr_pgs, u8 * vec, int level, u64 addr, u64 * entry)
 {
-    if (validate_virtual(addr, length)) {
-        u32 vlen = pad(length, PAGESIZE) >> PAGELOG;
-        // presumably it wants the right valid bits set? - go doesn't seem to use it this way
-        for (int i = 0; i< vlen; i++) vec[i] = 1;
-        return 0;
+    u64 e = *entry;
+    u64 pgoff, i;
+
+    if (pt_entry_is_present(e)) {
+        pgoff = (addr - base) >> PAGELOG;
+
+        if (pt_entry_is_fat(level, e)) {
+            /* whole level is mapped */
+            for (i = 0; i < 512 && (pgoff + i < nr_pgs); i++) {
+                vec[pgoff + i] = 1;
+	    }
+        } else if (pt_entry_is_pte(level, e)) {
+            vec[pgoff] = 1;
+        }
     }
-    return -ENOMEM;
+
+    return true;
+}
+
+static CLOSURE_0_1(mincore_vmap_gap, void, range);
+static void mincore_vmap_gap(range r)
+{
+    rprintf("mincore: found gap [0x%lx, 0x%lx)\n", r.start, r.end);
```

Comment:
This output should probably be debug-only. If you wish, you could modify rangemap_range_find_gaps() to accept an optional (maybe null) range_handler if no real callback on a gap is needed.